src/Interpreters/DatabaseCatalog.cpp

```
@@ -645,6 +652,47 @@ void DatabaseCatalog::updateDatabaseName(const String & old_name, const String &
     }
 }
 
+void DatabaseCatalog::updateDatabaseComment(const String & database_name, const String & old_database_comment)
+{
+    std::lock_guard lock{databases_mutex};
+    auto it = databases.find(database_name);
+    assert(it != databases.end());
+    auto db = it->second;
+    if (!db)
+        throw Exception(ErrorCodes::UNKNOWN_DATABASE, "Database {} does not exist", backQuoteIfNeed(database_name));
+
+    ASTPtr ast = db->getCreateDatabaseQuery();
+    if (!ast)
+        throw Exception(ErrorCodes::THERE_IS_NO_QUERY, "Unable to show the create query of database {}", backQuoteIfNeed(database_name));
+
+    auto * ast_create_query = ast->as<ASTCreateQuery>();
+    if (!ast_create_query->is_dictionary)
+        ast_create_query->attach = true;
+
+    WriteBufferFromOwnString statement_buf;
+    formatAST(*ast_create_query, statement_buf, false);
+    writeChar('\n', statement_buf);
+    String statement = statement_buf.str();
+
+    auto database_metadata_tmp_path = fs::path("metadata") / (escapeForFileName(database_name) + ".sql.tmp");
+    auto database_metadata_path = fs::path("metadata") / (escapeForFileName(database_name) + ".sql");
+    auto db_disk = getContext()->getDatabaseDisk();
+
+    writeMetadataFile(
+        db_disk, /*file_path=*/database_metadata_tmp_path, /*content=*/statement, getContext()->getSettingsRef()[Setting::fsync_metadata]);
+
+    try
+    {
+        /// rename atomically replaces the old file with the new one.
+        db_disk->replaceFile(database_metadata_tmp_path, database_metadata_path);
+    }
+    catch (...)
+    {
+        db_disk->removeFileIfExists(database_metadata_tmp_path);
+        db->setDatabaseComment(old_database_comment);
```

Do we need to re-throw the exception here?

---

Updating metadata,currently, writes to .tmp file before renaming it to final name. As far as I understand, it happens to prevent corrupted files and ensure atomic file writes.

With this approach, I think that we still need a try-catch in updateMetadataFile function for replcaing temporary file with final one.

---

Yes, you are right. We still need try-catch in the `updateMetadataFile`. However, it only handles metadata files, it doesn't need to take care of the database comment. So we can decouple the logic.